### PR TITLE
Bucket strategy managment and segment controller

### PR DESCRIPTION
**Pull request by hanahmily**
These changes tend to introduce a bucket-based strategy manager to move segments and blocks forward. The new manager enhancements the segment in this PR. The block's going to implement it in the following up commits.

Caveat: the manager doesn't clean up segments/blocks, which are not the current reporter. Another resource controller will take it according to the memory state. 

Signed-off-by: Gao Hongtao <hanahmily@gmail.com>

### CODE REVIEW

**Review by wu-sheng — DISMISSED** [review]

File: banyand/tsdb/bucket/strategy.go [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+		for s.current == nil {
+			s.current = s.ctrl.Current()
+		}
+		s.next = nil
+	}
+	reset()
+	go func(s *Strategy) {
+		var err error
+		for {
+		bucket:
```
Comment (lujiajing1126) [inline banyand/tsdb/bucket/strategy.go:95]: I suppose a single `for-select` loop is enough for this case

File: banyand/tsdb/bucket/strategy.go [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+			c := s.current.Report()
+			for {
+				select {
+				case status, more := <-c:
```
Comment (lujiajing1126) [inline banyand/tsdb/bucket/strategy.go:99]:
```suggestion
				case status, closed := <-c:
```

File: banyand/tsdb/bucket/strategy.go [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+						reset()
+						goto bucket
+					}
+					ratio := Ratio(status.Volume) / Ratio(status.Capacity)
+					if ratio >= s.ratio && s.next == nil {
+						s.next, err = s.ctrl.Next()
+						if errors.Is(err, ErrNoMoreBucket) {
+							return
+						}
+						if err != nil {
```
Comment (lujiajing1126) [inline banyand/tsdb/bucket/strategy.go:110]: if err is not nil, probably it is dangerous to proceed within the current loop

File: banyand/tsdb/shard.go [hunk 4]
```diff
@@ -52,66 +54,234 @@ func (s *shard) Index() IndexDatabase {
...
+	}
+	return sc.load(ctx, suffix, segPath)
+}
+
+func (sc *segmentController) load(ctx context.Context, suffix, path string) (seg *segment, err error) {
+	seg, err = openSegment(ctx, suffix, path, sc.intervalRule)
+	if err != nil {
+		return nil, err
+	}
+	{
```
Comment (lujiajing1126) [inline banyand/tsdb/shard.go:275]: unnecessary pair of braces?

File: banyand/tsdb/segment.go [hunk 5]
```diff
@@ -19,42 +19,74 @@ package tsdb
...
+	interval := s.Duration() >> 4
+	if interval < 100*time.Millisecond {
+		interval = 100 * time.Millisecond
 	}
-	return greaterAndEqualStart && s.endTime.After(ts)
+	go func() {
+		defer close(ch)
+		for {
+			ticker := time.NewTicker(interval)
+			defer ticker.Stop()
```
Comment (lujiajing1126) [inline banyand/tsdb/segment.go:54]: resource leak here

**Review by lujiajing1126 — CHANGES_REQUESTED** [review]

File: banyand/tsdb/segment.go [hunk 6]
```diff
@@ -19,42 +19,74 @@ package tsdb
...
+	TimeRange
 }
 
-func (s *segment) contains(ts time.Time) bool {
-	greaterAndEqualStart := s.startTime.Equal(ts) || s.startTime.Before(ts)
-	if s.endTime.IsZero() {
-		return greaterAndEqualStart
+func (s *segment) Report() bucket.Channel {
+	ch := make(bucket.Channel)
+	interval := s.Duration() >> 4
```
Comment (lujiajing1126) [inline banyand/tsdb/segment.go:46]: why we need to divide by 2^4?

File: banyand/tsdb/bucket/strategy.go (hunk 1 above)
Comment (hanahmily) [reply banyand/tsdb/bucket/strategy.go:95]: ` s.current.Report()` will return a new channel. We should reuse it by introducing an outer loop.

File: banyand/tsdb/bucket/strategy.go (hunk 3 above)
Comment (hanahmily) [reply banyand/tsdb/bucket/strategy.go:110]: The loop keeps calling the `next` continually till it successfully. This piece could fix accidental failures.

File: banyand/tsdb/shard.go (hunk 4 above)
Comment (hanahmily) [reply banyand/tsdb/shard.go:275]: it tends to reduce the scope of the lock

File: banyand/tsdb/segment.go (hunk 5 above)
Comment (hanahmily) [reply banyand/tsdb/segment.go]: done

File: banyand/tsdb/segment.go (hunk 6 above)
Comment (hanahmily) [reply banyand/tsdb/segment.go:46]: get rid of setting the interval by hand. The interval = the segment duration / 16

File: banyand/tsdb/bucket/strategy.go (hunk 1 above)
Comment (lujiajing1126) [reply banyand/tsdb/bucket/strategy.go:95]: But the outer `for-loop` is not necessary?

```go
var err error
bucket:
c := s.current.Report()
for {
  select { ... }
}
```

**Comment by lujiajing1126** [comment]
LGTM

**Review by lujiajing1126 — APPROVED** [review]

**Pull request merged**